File: slider/src/androidTest/java/com/adsamcik/slider/IntSliderInstrumentationTest.java
package com.adsamcik.slider;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Build;
import android.os.Looper;
import android.preference.PreferenceManager;
import android.support.test.InstrumentationRegistry;
import android.support.test.runner.AndroidJUnit4;
import android.widget.SeekBar;
import android.widget.TextView;

import com.adsamcik.slider.ScaleFunctions.LinearScale;

import junit.framework.Assert;

import org.junit.Test;
import org.junit.runner.RunWith;

import java.security.InvalidParameterException;
import java.util.concurrent.atomic.AtomicInteger;

import static org.junit.Assert.assertEquals;

/**
 * Instrumented test, which will execute on an Android device.
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */
@RunWith(AndroidJUnit4.class)
public class IntSliderInstrumentationTest {
	private Context appContext = InstrumentationRegistry.getTargetContext();

	@Test
	public void baseTest() throws Exception {
		IntSlider slider = new IntSlider(appContext);
		slider.setMinValue(1);
		slider.setMaxValue(9);
		slider.setStep(2);
		slider.setValue(5);
		slider.setTextView(null, null);

		assertEquals(1, (long) slider.getMinValue());
		assertEquals(9, (long) slider.getMaxValue());
		assertEquals(8, (long) slider.getMax());
		assertEquals(4, (long) slider.getProgress());
		assertEquals(5, (long) slider.getValue());

		Slider.IScale<Integer> scale = LinearScale.getIntegerScale();
		slider.setScale(scale);
		assertEquals(scale, slider.getScale());
	}

	@Test
	public void textViewTest() throws Exception {
		IntSlider slider = new IntSlider(appContext);
		TextView textView = new TextView(appContext);

		slider.setTextView(textView, value -> "Test " + value);
		slider.setMinValue(0);
		slider.setMaxValue(10);
		slider.setValue(5);

		assertEquals("Test 5", textView.getText());

		slider.setTextView(null, value -> "Test " + value);
		slider.setValue(4);
		assertEquals("Test 5", textView.getText());
		assertEquals(4, (long) slider.getValue());

		slider.setTextView(textView, value -> "Test " + value);
		slider.setTextView(textView, null);
		slider.setValue(6);
		assertEquals("Test 4", textView.getText());
		assertEquals(6, (long) slider.getValue());

		slider.setTextView(null, null);
		slider.setValue(8);
		assertEquals("Test 4", textView.getText());
		assertEquals(8, (long) slider.getValue());
	}

	@Test
	public void progressTest() throws Exception {
		if (Looper.myLooper() == null)
			Looper.prepare();

		IntSlider slider = new IntSlider(appContext);
		slider.setMinValue(0);
		slider.setMaxValue(10);
		slider.setValue(5);

		assertEquals(5, slider.getProgress());

		if (Build.VERSION.SDK_INT >= 24)
			slider.setValue(8, false);
		else
			slider.setProgress(8);

		assertEquals(8, slider.getProgress());

		if (Build.VERSION.SDK_INT >= 24)
			slider.setValue(3, true);
		else
			slider.setProgress(3);

		assertEquals(3, slider.getProgress());
	}

	@Test
	public void moveTest() throws Exception {
		IntSlider slider = new IntSlider(appContext);
		slider.setMinValue(0);
		slider.setMaxValue(12);
		slider.setStep(3);
		for (int i = 0; i < 12; i++) {
			slider.setValue(i);
			assertEquals(Math.round(i / 3.0) * 3, (long) slider.getValue());
		}
	}

	@Test
	public void negativeNumberTest() throws Exception {
		IntSlider slider = new IntSlider(appContext);
		slider.setMinValue(-15);
		slider.setMaxValue(15);
		slider.setStep(3);
		for (int i = -15; i <= 15; i++) {
			slider.setValue(i);
			assertEquals(Math.round(i / 3.0) * 3, (long) slider.getValue());
		}
	}

	@Test
	public void stepTest() throws Exception {
		IntSlider slider = new IntSlider(appContext);
		slider.setMinValue(0);
		slider.setMaxValue(10);

		slider.setValue(5);

		Assert.assertEquals(5, (int) slider.getValue());

		slider.setStep(2);

		int value = slider.getValue();
		Assert.assertTrue(value == 4 || value == 6);

		slider.setStep(3);
		slider.setValue(5);
		Assert.assertEquals(6, (int) slider.getValue());
	}

	@Test
	public void sharedPreferences() throws Exception {
		final SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(appContext);
		final String prefName = "TESTING PREFERENCE";

		preferences.edit().remove(prefName).apply();

		IntSlider slider = new IntSlider(appContext);
		slider.setMinValue(-5);
		slider.setMaxValue(5);
		slider.setStep(2);

		slider.setPreferencesAndLoad(preferences, prefName, 1);

		Assert.assertEquals(1, (long) slider.getValue());

		slider.setValue(4);

		int value = slider.getValue();
		Assert.assertEquals(value, preferences.getInt(prefName, Integer.MIN_VALUE));

		slider.setPreferences(null, null);

		slider.setValue(1);

		Assert.assertEquals(value, preferences.getInt(prefName, Integer.MIN_VALUE));
		Assert.assertEquals(1, (int) slider.getValue());

		//cleanup
		preferences.edit().remove(prefName).apply();
	}

	private AtomicInteger atomicInteger = new AtomicInteger(0);

	@Test
	public void callbackTests() throws Exception {
		IntSlider slider = new IntSlider(appContext);

		Slider.OnValueChangeListener<Integer> valueChangeListener = (value, fromUser) -> {
			atomicInteger.incrementAndGet();
		};

		slider.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
			@Override
			public void onProgressChanged(SeekBar seekBar, int i, boolean b) {
				atomicInteger.incrementAndGet();
			}

			@Override
			public void onStartTrackingTouch(SeekBar seekBar) {

			}

			@Override
			public void onStopTrackingTouch(SeekBar seekBar) {

			}
		});

		slider.setOnValueChangeListener(valueChangeListener);
		slider.setValue(5);
		assertEquals(2, atomicInteger.get());

		slider.setOnValueChangeListener(null);
		slider.setOnSeekBarChangeListener(null);
		slider.setValue(3);
		assertEquals(2, atomicInteger.get());
	}

	@Test
	public void exceptionTest() {
		IntSlider slider = new IntSlider(appContext);
		EAssert.assertException(() -> slider.setSliderStep(-5), RuntimeException.class);

		EAssert.assertException(() -> slider.setMinValue(slider.getMaxValue()), InvalidParameterException.class);
		EAssert.assertException(() -> slider.setMaxValue(slider.getMinValue()), InvalidParameterException.class);
	}
}

File: slider/src/androidTest/java/com/adsamcik/slider/IntSliderInstrumentationTest.kt
package com.adsamcik.slider

import android.content.Context
import android.os.Looper
import androidx.preference.PreferenceManager
import androidx.test.core.app.ApplicationProvider
import com.adsamcik.slider.abstracts.Slider
import com.adsamcik.slider.abstracts.SliderExtension
import com.adsamcik.slider.extensions.IntSliderSharedPreferencesExtension
import com.adsamcik.slider.implementations.IntSlider
import com.adsamcik.slider.scaleFunctions.LinearScale
import org.junit.Assert
import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.concurrent.atomic.AtomicInteger
import kotlin.math.roundToLong

/**
 * Instrumented test, which will execute on an Android device.
 *
 * @see [Testing documentation](http://d.android.com/tools/testing)
 */
class IntSliderInstrumentationTest {
	companion object {
		private const val FLUID_DELTA = 0.0001f
	}

	private val appContext = ApplicationProvider.getApplicationContext<Context>()

	private val atomicInteger = AtomicInteger(0)


	@Test
	@Throws(Exception::class)
	fun baseTest() {
		val slider = IntSlider(appContext)
		slider.minValue = 1
		slider.maxValue = 9
		slider.step = 2
		slider.value = 5
		slider.setLabelFormatter(null)

		assertEquals(1, slider.minValue.toLong())
		assertEquals(9, slider.maxValue.toLong())
		assertEquals(0.5f, slider.fluidPosition, FLUID_DELTA)
		assertEquals(5, slider.value.toLong())

		val scale = LinearScale.integerScale
		slider.scale = scale
		assertEquals(scale, slider.scale)
	}

	@Test
			/**
			 * This tests check if description is set properly.
			 * This also ensures updating description does not crash the slider.
			 */
	fun descriptionTest() {
		val slider = IntSlider(appContext)
		slider.minValue = 1
		slider.maxValue = 9
		slider.step = 2
		slider.value = 5
		val description = "TestDescription"
		slider.description = description

		assertEquals(description, slider.description)

		val scale = LinearScale.integerScale
		slider.scale = scale
		assertEquals(scale, slider.scale)
	}

	@Test
	@Throws(Exception::class)
	fun progressTest() {
		if (Looper.myLooper() == null)
			Looper.prepare()

		val slider = IntSlider(appContext)
		slider.minValue = 0
		slider.maxValue = 10
		slider.value = 5

		assertEquals(0.5f, slider.fluidPosition, FLUID_DELTA)

		slider.value = 8

		assertEquals(0.8f, slider.fluidPosition, FLUID_DELTA)


		slider.value = 3

		assertEquals(0.3f, slider.fluidPosition, FLUID_DELTA)
	}

	@Test
	@Throws(Exception::class)
	fun moveTest() {
		val slider = IntSlider(appContext)
		slider.minValue = 0
		slider.maxValue = 12
		slider.step = 3
		for (i in 0..11) {
			slider.value = i
			assertEquals((i / 3.0).roundToLong() * 3, slider.value.toLong())
		}
	}

	@Test
	@Throws(Exception::class)
	fun negativeNumberTest() {
		val slider = IntSlider(appContext)
		slider.minValue = -15
		slider.maxValue = 15
		slider.step = 3
		for (i in -15..15) {
			slider.value = i
			assertEquals((i / 3.0).roundToLong() * 3, slider.value.toLong())
		}
	}

	@Test
	@Throws(Exception::class)
	fun stepTest() {
		val slider = IntSlider(appContext)
		slider.minValue = 0
		slider.maxValue = 10

		slider.value = 5

		assertEquals(5, slider.value)

		slider.step = 2

		val value = slider.value
		Assert.assertTrue(value == 4 || value == 6)

		slider.step = 3
		slider.value = 5
		assertEquals(6, slider.value)
	}

	@Test
	@Throws(Exception::class)
	fun sharedPreferences() {
		val preferences = PreferenceManager.getDefaultSharedPreferences(appContext)
		val prefName = "TESTING PREFERENCE"

		preferences.edit().clear().apply()

		val slider = IntSlider(appContext)
		slider.minValue = -5
		slider.maxValue = 5
		slider.step = 2

		val extension = IntSliderSharedPreferencesExtension(preferences, prefName, 1)
		slider.addExtension(extension)

		assertEquals(1, slider.value.toLong())

		slider.value = 4

		val value = slider.value
		assertEquals(value, preferences.getInt(prefName, Integer.MIN_VALUE))

		slider.removeExtension(extension)

		slider.value = 1

		assertEquals(value, preferences.getInt(prefName, Integer.MIN_VALUE))
		assertEquals(1, slider.value)

		//cleanup
		preferences.edit().remove(prefName).apply()
	}

	@Test
	@Throws(Exception::class)
	fun callbackTests() {
		val slider = IntSlider(appContext)

		val lastValue = 5

		val extension = object : SliderExtension<Int> {
			override fun onValueChanged(
					slider: Slider<Int>,
					value: Int,
					position: Float,
					isFromUser: Boolean
			) {
				assertEquals(lastValue, value)
				assertEquals(false, isFromUser)
				atomicInteger.incrementAndGet()
			}
		}

		slider.addExtension(extension)
		slider.value = lastValue
		assertEquals(1, atomicInteger.get().toLong())

		slider.removeExtension(extension)
		slider.value = 3
		assertEquals(1, atomicInteger.get().toLong())
	}

	@Test
	fun exceptionTest() {
		val slider = IntSlider(appContext)
		AssertUtility.assertException(
				{ slider.step = -5 },
				IllegalArgumentException::class.java
		)

		AssertUtility.assertException(
				{ slider.minValue = slider.maxValue },
				IllegalArgumentException::class.java
		)
		AssertUtility.assertException(
				{ slider.maxValue = slider.minValue },
				IllegalArgumentException::class.java
		)
	}
}
